Improve Porsche Vehicle Support for PHEV

DerAndereAndi opened the pull request:
Only the Taycan supports the emobility API, all other PHEV Porsche vehicles don’t. So adopt this to support both scenarios.

- Check if the emobility API token is available
- Check if the given car is a Taycan and only use the emobility API for this car

Caution: This is only tested with a Taycan, due to lacking access to a Porsche PHEV vehicle.

DerAndereAndi committed:
Improve Porsche Vehicle Support for PHEV

Only the Taycan supports the emobility API, all other PHEV Porsche vehicles don’t. So adopt this to support both scenarios.

- Check if the emobility API token is available
- Check if the given car is a Taycan and only use the emobility API for this car

Caution: This is only tested with a Taycan, due to lacking access to a Porsche PHEV vehicle.

andig commented:
Spätestens bei 500 Zeilen Code wäre es vielleicht gut, die 2 Porsche APIs in 2 Dateien in einem porsche Subfolder auszulagern?

DerAndereAndi commented:
Die `ChargeState` API ist unterschiedlich, der Rest nicht. Hast ein Beispiel wie du so was schon gemacht hast?

andig commented:
Mhhm. Ich hab immer versucht, die Tokens von oauth zu nutzen, da ist auch Validity mit drin und du bekommst nen http client dafür. Ohne Refresh ists aber nur die halbe Miete 😔

andig commented:
Dann die Api separat kapseln? Ich hab immer api, identity (=Token) und provider (=evcc Integration) unterschieden, z.b. bei VW.

DerAndereAndi committed:
Restructure code

Used PSA implementation as an orientation

DerAndereAndi committed:
Restructure and separate Porsche APIs

This cleans up the code a bit and separates the server side available APIs for the different car types.

DerAndereAndi commented:
@andig Thanks a ton helping getting the restructured code work. Couldn't have done it myself.

DerAndereAndi committed:
Added missing file

DerAndereAndi committed:
Fixed range request not working

DerAndereAndi committed:
Cleanup namings a bit

andig committed:
Remove clientid

DerAndereAndi committed:
Cleanup

- Fixed  typos
- Fixed namings
- Removed unnecessary variables

DerAndereAndi committed:
More cleanup

- Use oauth2.Token struct for the token data
- More namings cleanup
- Removed more unnecessary variables

DerAndereAndi committed:
More restructuring

- Rename API to Identity
- Move request helper method into Providers
- Provide the Providers with the tokens they need
- Various naming adjustments
